src/python/pants/core/goals/test.py

```diff
@@ -869,6 +887,14 @@ async def _run_debug_tests(
     return Test(exit_code)
 
 
+def save_test_result_source_report_file(run_id: RunId, results: dict[str, str]) -> None:
+    """Save a JSON file with the information about the source of test result."""
+    timestamp = datetime.now().strftime("%Y_%m_%d_%H_%M_%S")
+    obj = json.dumps({"timestamp": timestamp, "run_id": run_id, "sources": results})
+    with safe_open(f"test_result_source_report_runid{run_id}_{timestamp}.json", "w") as fh:
```

Comment:
This should be concurrent-safe since every invocation of the test goal would have a unique id, so multiple files can be created in parallel with no racing. This follows closely the naming pattern of the workunit-logger files, e.g. `pants_run_2024_05_11_23_39_48_438_cf61f6cd0266404bb307a9ff3103cea4`. I don't think a GUID is justified here and run id should suffice.

Also relying on run id provides a hint at the order in which the goals were run (in case there are multiple `test` goal invocations as part of the build).

After running the `test` goal a few times one after another:

```
$ ls test_result* -1
test_result_source_report_runid7_2024_07_14_23_30_02.json
test_result_source_report_runid8_2024_07_14_23_30_12.json
test_result_source_report_runid9_2024_07_14_23_30_17.json
```

Reply:
An example output of the file:

```
{"timestamp": "2024_07_14_23_30_02", "run_id": 7, "sources": {"src/python/pants/core/goals/test_test.py:tests": "memoized"}}
```

Reply:
Another one:

```
{"timestamp": "2024_07_14_23_34_03", "run_id": 0, "sources": {"helloworld/greet/greeting_test.py:tests": "hit_locally", "helloworld/translator/translator_test.py:tests": "hit_locally"}}
```